File: fathersDay/Dsa/Day 31 ( 6. Zigzag Conversion)/Solution.py
class Solution:
    def ZigZagConvert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s
        
        # Create a list of strings for each row
        rows = [''] * numRows
        current_row = 0
        going_down = False
        
        for char in s:
            rows[current_row] += char
            
            # Change direction if we hit the top or bottom row
            if current_row == 0:
                going_down = True
            elif current_row == numRows - 1:
                going_down = False
            
            # Move to the next row
            current_row += 1 if going_down else -1
        
        # Join all rows to get the final zigzag string
        return ''.join(rows)
# Example usage:
solution = Solution()
s = "PAYPALISHIRING"
numRows = 3
result = solution.ZigZagConvert(s, numRows)
print(result)  # Output: "PAHNAPLSIIGYIR"

File: fathersDay/Dsa/Day 31 ( 6. Zigzag Conversion)/Solution.cpp
#include <iostream>
#include <vector>
#include <string>

class Solution {
public:
    std::string ZigZagConvert(std::string s, int numRows) {
        if (numRows == 1 || numRows >= s.length()) {
            return s;
        }

        std::vector<std::string> rows(numRows);
        int current_row = 0;
        bool going_down = false;

        for (char c : s) {
            rows[current_row] += c;
            if (current_row == 0 || current_row == numRows - 1) {
                going_down = !going_down;
            }
            current_row += going_down ? 1 : -1;
        }

        std::string result;
        for (const std::string& row : rows) {
            result += row;
        }
        return result;
    }
};

int main() {
    Solution solution;
    std::string s = "PAYPALISHIRING";
    int numRows = 3;
    std::string result = solution.ZigZagConvert(s, numRows);
    std::cout << result << std::endl; // Output: "PAHNAPLSIIGYIR"
    return 0;
}
